packages/itertools/iter.pony

```diff
@@ -369,12 +368,34 @@ class Iter[A] is Iterator[A]
     ## Example
 
     ```pony
-    Iter[I64]([as I64: 1; 1; 2; 3; 3; 2; 2].values())
+    Iter[USize]([as USize: 1; 1; 2; 3; 3; 2; 2].values())
       .dedup()
     ```
     `1 2 3 2`
     """
-  */
+    Iter[A!](
+      object is Iterator[A!]
+        var _prev_hash: (USize | None) = None
+
+        fun ref has_next(): Bool =>
+          _iter.has_next()
+
+        fun ref next(): A! ? =>
+          let v = _iter.next()?
+          let cur_hash = H.hash(v)
+          match _prev_hash
+          | let x: USize =>
+            if x == cur_hash then
```

Comment:
To me it seems somewhat "off" that we are only using `H.hash` and not also using `H.eq` to confirm true value equality (instead of just hash-equivalence).

Some hash functions may have a non-negligible collision rate for non-identical elements, and because `HashFunction` has both a `hash` method and an `eq` method, users may be relying/assuming that the `H.eq` function will be used to check for true value equivalence, and be resilient to false collisions.

Reply:
I had to look at the source again around HashFunction, but I see what you mean and will make the necessary change(s) to use H.eq where needed.

Reply:
@jemc Please check this again and make suggestions for improvement. In order to use `H.eq` I needed to keep both the previous hash and previous value. I initially tried to solve this with a `var _prev: ((A!, USize) | None) = None` but unpacking the tuple was causing me issues with ponyc insisting `_iter.next()?` having a return type of `(A #any ! | None val^)` so I split the value and hash into separate variables. I presume based on your comment here and the documentation around `H.eq` that the order of equality should be checking the hash match then using `H.eq` but am not 100% certain it should not be the other way around.

Reply:
I also think this code is a bit messy right now with `cur_{value,hash}`, `_prev_{value,hash}`, and `prev_value` so if you have advice on making these names less redundant or the code less repetitive, please let me know!